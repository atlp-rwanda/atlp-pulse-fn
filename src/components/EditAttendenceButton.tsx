import React, { useEffect, useState } from 'react';
import AttendanceSymbols from './AttendanceSymbols';
import { TraineeAttendanceDataInterface } from '../pages/TraineeAttendanceTracker';

interface EditAttendanceProps {
  setTraineeAttendanceData: React.Dispatch<React.SetStateAction<any[]>>;
  setUpdated: React.Dispatch<React.SetStateAction<boolean>>;
  week: string;
  day: string;
  phase: string;
  traineeId: string;
}

function EditAttendanceButton({
  week,
  day,
  phase,
  traineeId,
  setTraineeAttendanceData,
  setUpdated,
}: EditAttendanceProps) {
  const [openEdit, setOpenEdit] = useState<boolean>(false);

  useEffect(() => {
    setOpenEdit(false);
  }, [week, phase, day]);

  const handleUpdateAttendance = (score: string) => {
    setTraineeAttendanceData((prev) =>
      prev.map((attendanceData) => {
        if (attendanceData.week === week && attendanceData.phase === phase) {
          const updatedDay = attendanceData.days[
            day as keyof typeof attendanceData.days
          ].map((traineeData: TraineeAttendanceDataInterface) => {
            if (
              traineeData.trainee.id === traineeId &&
              traineeData.status.toString() !== score.toString()
            ) {
              setUpdated(true);
              return {
                trainee: traineeData.trainee,
                status: score,
              };
            }
            return traineeData;
          });

          return {
            ...attendanceData,
            days: { ...attendanceData.days, [day]: updatedDay },
          };
        }
        return attendanceData;
      }),
    );
    setOpenEdit(false);
  };

  return (
    <div className="inline-block cursor-pointer relative">
      {!openEdit && (
        <span
          onClick={() => setOpenEdit(true)}
          className="px-2 py-[2px] border dark:border-white border-black rounded-md text-[.85rem]"
          data-testid="edit-button"
        >
          Edit
        </span>
      )}
      {openEdit && (
        <div className="flex items-center gap-1 xmd:gap-2">
          <div
            onClick={() => handleUpdateAttendance('2')}
            data-testid="score-2"
          >
            <AttendanceSymbols status={2} />
          </div>
          <div
            onClick={() => handleUpdateAttendance('1')}
            data-testid="score-1"
          >
            <AttendanceSymbols status={1} />
          </div>
          <div
            onClick={() => handleUpdateAttendance('0')}
            data-testid="score-0"
          >
            <AttendanceSymbols status={0} />
          </div>
        </div>
      )}
    </div>
  );
}

export default EditAttendanceButton;
