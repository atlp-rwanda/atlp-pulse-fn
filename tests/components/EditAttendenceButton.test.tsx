import React, { useState } from 'react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import EditAttendenceButton from '../../src/components/EditAttendenceButton';

const initialData = [
  {
    week: '1',
    phase: 'Phase I',
    days: {
      mon: [
        {
          trainee: {
            id: 'trainee-id-1234',
          },
          status: '1',
        },
      ],
    },
  },
];
const initialData2 = [
  {
    week: '2',
    phase: 'Phase I',
    days: {
      mon: [
        {
          trainee: {
            id: 'trainee-id-1234',
          },
          status: '0',
        },
      ],
    },
  },
];

describe('attendance update button', () => {
  it('Renders attendance update component', () => {
    const elem = renderer
      .create(
        <MemoryRouter>
          <EditAttendenceButton
            day="mon"
            week={3}
            traineeId="trainee-id-1234"
            setUpdated={(value) => value}
            phase="Phase !"
            setTraineeAttendanceData={() => true}
          />
        </MemoryRouter>,
      )
      .toJSON();
    expect(elem).toMatchSnapshot();
  });

  it('Trigger Edit button and update to score 2', () => {
    
    const setTraineeAttendanceDataMock = jest.fn();
    const setUpdatedMock = jest.fn();
    setTraineeAttendanceDataMock.mockImplementation((update) => update(initialData))


    render(
      <MemoryRouter>
        <EditAttendenceButton
          day="mon"
          week={1}
          traineeId="trainee-id-1234"
          setUpdated={setUpdatedMock}
          phase="Phase I"
          setTraineeAttendanceData={setTraineeAttendanceDataMock}
        />
      </MemoryRouter>,
    );

    const editButton = screen.getByText('Edit');
    expect(editButton).toBeInTheDocument();

    fireEvent.click(editButton);
    fireEvent.click(screen.getByTestId('score-2'));
  });
  it('Trigger Edit button and update to score 1', () => {
    
    const setTraineeAttendanceDataMock = jest.fn();
    const setUpdatedMock = jest.fn();
    setTraineeAttendanceDataMock.mockImplementation((update) => update(initialData))


    render(
      <MemoryRouter>
        <EditAttendenceButton
          day="mon"
          week={1}
          traineeId="trainee-id-1234"
          setUpdated={setUpdatedMock}
          phase="Phase I"
          setTraineeAttendanceData={setTraineeAttendanceDataMock}
        />
      </MemoryRouter>,
    );

    const editButton = screen.getByText('Edit');
    expect(editButton).toBeInTheDocument();

    fireEvent.click(editButton);
    fireEvent.click(screen.getByTestId('score-1'));
  });
  it('Trigger Edit button and update to score 0', () => {
    
    const setTraineeAttendanceDataMock = jest.fn();
    const setUpdatedMock = jest.fn();
    setTraineeAttendanceDataMock.mockImplementation((update) => update(initialData2))


    render(
      <MemoryRouter>
        <EditAttendenceButton
          day="mon"
          week={1}
          traineeId="trainee-id-1234"
          setUpdated={setUpdatedMock}
          phase="Phase I"
          setTraineeAttendanceData={setTraineeAttendanceDataMock}
        />
      </MemoryRouter>,
    );

    const editButton = screen.getByText('Edit');
    expect(editButton).toBeInTheDocument();

    fireEvent.click(editButton);
    fireEvent.click(screen.getByTestId('score-0'));
  });
});
