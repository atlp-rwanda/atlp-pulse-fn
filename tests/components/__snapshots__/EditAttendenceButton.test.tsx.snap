// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attendance update button Renders attendance update component 1`] = `
<div
  className="inline-block cursor-pointer relative"
>
  <span
    className="px-2 py-[2px] border dark:border-white border-black rounded-md text-[.85rem]"
    data-testid="edit-button"
    onClick={[Function]}
  >
    Edit
  </span>
</div>
`;
